File: app/src/main/java/com/fr1014/mycoludmusic/ui/vm/CommonViewModel.java
package com.fr1014.mycoludmusic.ui.vm;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.fr1014.mycoludmusic.data.DataRepository;
import com.fr1014.mycoludmusic.data.entity.http.wangyiyun.playlist.PlayListDetailEntity;
import com.fr1014.mycoludmusic.data.entity.http.wangyiyun.playlist.Playlist;
import com.fr1014.mycoludmusic.data.entity.http.wangyiyun.playlist.WYUserPlayList;
import com.fr1014.mycoludmusic.data.entity.http.wangyiyun.playlist.TrackIds;
import com.fr1014.mycoludmusic.data.entity.http.wangyiyun.user.Profile;
import com.fr1014.mycoludmusic.data.entity.http.wangyiyun.user.WYLikeIdList;
import com.fr1014.mycoludmusic.data.source.local.room.MusicLike;
import com.fr1014.mycoludmusic.musicmanager.Preferences;
import com.fr1014.mycoludmusic.rx.RxSchedulers;
import com.fr1014.mymvvm.base.BaseViewModel;
import com.fr1014.mymvvm.base.BusLiveData;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.functions.Consumer;

public class CommonViewModel extends BaseViewModel<DataRepository> {

    protected BusLiveData<Long[]> playListDetailIds;
    protected MutableLiveData<List<Playlist>> playlistWYLive;
    private MutableLiveData<PlayListDetailEntity> playListDetailLive;

    public CommonViewModel(@NonNull Application application) {
        super(application);
    }

    public CommonViewModel(@NonNull Application application, DataRepository model) {
        super(application, model);
    }

    public LiveData<PlayListDetailEntity> getPlayListDetailInfo() {
        if (playListDetailLive == null) {
            playListDetailLive = new MutableLiveData<>();
        }
        return playListDetailLive;
    }

    public LiveData<List<Playlist>> getPlaylistWYLive() {
        if (playlistWYLive == null) {
            playlistWYLive = new MutableLiveData<>();
        }
        return playlistWYLive;
    }

    public LiveData<Long[]> getPlayListDetail(long id) {
        if (playListDetailIds == null) {
            playListDetailIds = new BusLiveData<>();
        }
        getPlayListDetailEntity(id);
        return playListDetailIds;
    }

    private void getPlayListDetailEntity(final long id) {
        addSubscribe(model.getPlayListDetail(id)
                .compose(RxSchedulers.apply())
                .subscribe(new Consumer<PlayListDetailEntity>() {
                    @Override
                    public void accept(PlayListDetailEntity playListDetailEntity) throws Exception {
                        List<TrackIds> tracks = playListDetailEntity.playlist.getTrackIds();
                        Long[] ids = new Long[tracks.size()];
                        for (int index = 0; index < tracks.size(); index++) {
                            ids[index] = tracks.get(index).getId();
                        }
                        playListDetailLive.postValue(playListDetailEntity);
                        playListDetailIds.postValue(ids);
                    }
                }));
    }

    public void getWYUserPlayList() {
        Profile profile = Preferences.getUserProfile();
        if (profile == null) return;
        final long userId = profile.getUserId();
        if (userId != 0L)
            addSubscribe(
                    model.getWYUserPlayList(userId, String.valueOf(System.currentTimeMillis()))
                            .compose(RxSchedulers.apply())
                            .subscribe(new Consumer<WYUserPlayList>() {
                                @Override
                                public void accept(WYUserPlayList wyUserPlayList) throws Exception {
                                    playlistWYLive.postValue(wyUserPlayList.getPlaylist());
                                }
                            })
            );
    }

    public void getLikeIdList(Long uid) {
        addSubscribe(model.getWYLikeIdList(uid, String.valueOf(System.currentTimeMillis()))
                .compose(RxSchedulers.applyIO())
                .subscribe(new Consumer<WYLikeIdList>() {
                    @Override
                    public void accept(WYLikeIdList wyLikeIdList) throws Exception {
                        model.deleteAllLikeIds();
                        List<Long> ids = wyLikeIdList.getIds();
                        List<MusicLike> musicLikes = new ArrayList<>();
                        for (Long id : ids) {
                            musicLikes.add(new MusicLike(id));
                        }
                        model.insertAllLikeIds(musicLikes);
                    }
                })
        );
    }
}

File: app/src/main/java/com/fr1014/mycoludmusic/ui/vm/CommonViewModel.kt
package com.fr1014.mycoludmusic.ui.vm

import android.app.Application
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.rxjava2.cachedIn
import androidx.paging.rxjava2.observable
import com.fr1014.mycoludmusic.data.DataRepository
import com.fr1014.mycoludmusic.data.entity.http.wangyiyun.mv.MVData
import com.fr1014.mycoludmusic.data.entity.http.wangyiyun.playlist.PlayListDetailEntity
import com.fr1014.mycoludmusic.data.entity.http.wangyiyun.playlist.Playlist
import com.fr1014.mycoludmusic.data.entity.http.wangyiyun.song.SongsBean
import com.fr1014.mycoludmusic.data.source.local.room.MusicLike
import com.fr1014.mycoludmusic.musicmanager.Preferences
import com.fr1014.mycoludmusic.rx.RxSchedulers
import com.fr1014.mycoludmusic.ui.home.comment.paging3.CommentPagingSource
import com.fr1014.mymvvm.base.BaseViewModel
import com.fr1014.mymvvm.base.BusLiveData
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.Consumer
import io.reactivex.schedulers.Schedulers
import kotlin.collections.ArrayList

open class CommonViewModel : BaseViewModel<DataRepository> {

    var commentPagingSource: CommentPagingSource? = null

    private val mvDataLive: MutableLiveData<MVData> by lazy {
        MutableLiveData()
    }

    private val songInfoLive: MutableLiveData<SongsBean> by lazy {
        MutableLiveData()
    }

    private val playListDetailIds: BusLiveData<Array<Long>> by lazy {
        BusLiveData()
    }

    private val playlistWYLive: MutableLiveData<List<Playlist>> by lazy {
        MutableLiveData()
    }

    private val playListDetailLive: MutableLiveData<PlayListDetailEntity> by lazy {
        MutableLiveData()
    }

    constructor(application: Application) : super(application) {}
    constructor(application: Application, model: DataRepository?) : super(application, model) {}

    fun commentList(type: Int, id: Long, pageSize: Int) = Pager(PagingConfig(pageSize = 6)) {
        commentPagingSource = CommentPagingSource(type, id, pageSize)
        commentPagingSource!!
    }.observable.cachedIn(viewModelScope)

    val mvData: LiveData<MVData> = mvDataLive

    val songInfo: LiveData<SongsBean> = songInfoLive

    val playListDetailInfo: LiveData<PlayListDetailEntity> = playListDetailLive

    val playlistWYInfo: LiveData<List<Playlist>> = playlistWYLive

    fun getPlayListDetail(id: Long): LiveData<Array<Long>> {
        getPlayListDetailEntity(id)
        return playListDetailIds
    }

    /*
                dataRepository.getWYSongDetail(music.getId() + "")
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .map(songDetailEntity -> {
                        if (songDetailEntity.getSongs() != null && songDetailEntity.getSongs().size() > 0) {
                            music.setImgUrl(songDetailEntity.getSongs().get(0).getAl().getPicUrl());
                            music.setDuration(songDetailEntity.getSongs().get(0).getDt());
                        }
                        return music;
                    })
                    .subscribe(new SingleObserver<Music>() {
                        @Override
                        public void onSubscribe(@NonNull Disposable d) {

                        }

                        @Override
                        public void onSuccess(@NonNull Music music) {
                            CoverLoadUtils.get().loadRemoteCover(music);
                        }

                        @Override
                        public void onError(@NonNull Throwable e) {

                        }
                    });
     */

    fun getWYMVInfo(id: Long) {
        addSubscribe(model.getWYMVInfo(id)
                .compose(RxSchedulers.applyIO())
                .subscribe {
                    mvDataLive.postValue(it.data)
                })
    }

    fun getWYSongInfo(id: Long) {
        addSubscribe(model.getWYSongDetail(id.toString())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(Consumer {
                    songInfoLive.postValue(it.songs[0])
                })
        )
    }

    private fun getPlayListDetailEntity(id: Long) {
        addSubscribe(model.getPlayListDetail(id)
                .compose(RxSchedulers.apply())
                .subscribe { playListDetailEntity ->
                    val tracks = playListDetailEntity.playlist.trackIds
                    val ids = ArrayList<Long>()
                    for (index in tracks.indices) {
                        ids.add(tracks[index].id)
                    }
                    val array: Array<Long> = ids.toTypedArray()
                    playListDetailLive.postValue(playListDetailEntity)
                    playListDetailIds.postValue(array)
                })
    }

    fun getWYUserPlayList() {
        val userId = Preferences.getUserProfile()?.userId ?: return
        addSubscribe(model
                .getWYUserPlayList(userId, System.currentTimeMillis().toString())
                .compose(RxSchedulers.apply())
                .subscribe {
                    playlistWYLive.postValue(it.playlist)
                }
        )
    }

    fun getLikeIdList(uid: Long?) {
        addSubscribe(model!!.getWYLikeIdList(uid, System.currentTimeMillis().toString())
                .compose(RxSchedulers.applyIO())
                .subscribe { (_, _, ids) ->
                    model!!.deleteAllLikeIds()
                    val musicLikes: MutableList<MusicLike> = ArrayList()
                    for (id in ids) {
                        musicLikes.add(MusicLike(id))
                    }
                    model!!.insertAllLikeIds(musicLikes)
                }
        )
    }
}